### DISCUSSION TITLE
Livewire $listeners change from protected to public - is it OK?

### DISCUSSION THREAD

**Discussion by akulmehta** [livewire/livewire#7970] [1. Help] [answered]
I have a component where I would like to set a listener dynamically based on an event passed in. 

I have the following in the mount:

```php
   public $dispatchDeleteEvent = 'delete';
  

   protected $listeners = [];
   .
   .
   .
   public function mount()
    {
        $this->listeners[$this->dispatchDeleteEvent] = 'delete';
    }
```

The above doesn't work, but if I change the `$listeners` to public it works. 

Is it OK to do this or is this a security concern?

**Comment by danie-ramdhani** [comment]
idk if you're talking about V2 or V3. but, if it V3, there's `#[On(...)]` attributes. why not use it?
if it's v2, i wouldn't recommend do that. because if the function/event doesn't exist. it didn't do anything (that will make user confused is this works or not?).

EDIT:

oh, if you're talking about security. as long your user is authorized. it's ok (more in: https://livewire.laravel.com/docs/security).

**Reply by akulmehta** [reply]
@danie-ramdhani how would you use the `#[On(...)]` attribute in a dynamic way? I am talking about v3?

**Comment by joshhanley** [comment, marked as answer]
@akulmehta if you want to dynamically generate a listener, you should overwrite the `getListeners()` method, as making listeners public and doing it in mount isn't a great idea.

Note: dynamic listeners can only be generated the first time the component is loaded, once loaded they can't change.

```php
protected function getListeners()
{
    return [
        $this->dispatchDeleteEvent => 'delete',
    ];
}
```

Hope this helps!

**Reply by akulmehta** [reply]
Thanks - this works!

**Reply by joshhanley** [reply]
@akulmehta No worries 🙂
